doc/design/concurrent_programming.md

```diff
@@ -0,0 +1,163 @@
+# Design Doc: Concurrent Programming with Fluid
+
```

Comment:
I understand the `parallel_do` design and it is great.

But for curiosity, Is this design a concurrent programming use case just for data distribution?

or it includes the future design for the whole concurrent programming support? 

If the whole design, there needs some other concepts in addition to `parallel_for`, such as `waiting_for` or the low-level `mutex` and `condition_variable`.

And I wonder what is the capacity of fluid's concurrent programming? fully support the low-level concurrent concepts(very flexible) such as `mutex` or follow some classical framework(easy to use) such as `map/reduce`?

Reply:
Layer is the user-defined function, it looks like we need a `map` high order function to do parallel training and inference.